devtools/sendit/stream.go

```diff
@@ -0,0 +1,189 @@
+package sendit
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"sync"
+)
+
+// Stream is a ReadWriteCloser that can have it's read/write pipe replaced safely while actively sending data.
+type Stream struct {
+	readCh    chan io.ReadCloser
+	readEOFCh chan io.ReadCloser
+	writeCh   chan io.WriteCloser
+	closeCh   chan struct{}
+	mx        sync.Mutex
+
+	readyCh chan struct{}
+
+	isClosed bool
+	isReady  bool
+}
+
+// NewStream initializes a new stream. SetPipe must be called before data can be transferred.
+func NewStream() *Stream {
+	s := &Stream{
+		readCh:    make(chan io.ReadCloser, 1),
+		readEOFCh: make(chan io.ReadCloser, 1),
+		writeCh:   make(chan io.WriteCloser, 1),
+		closeCh:   make(chan struct{}),
+		readyCh:   make(chan struct{}),
+	}
+	return s
+}
+
+// Ready will return a channel indicating when the stream is ready for reading and writing.
+func (s *Stream) Ready() <-chan struct{} { return s.readyCh }
+
+func (s *Stream) Write(p []byte) (int, error) {
+	var w io.WriteCloser
+	select {
+	case w = <-s.writeCh:
+	case <-s.closeCh:
+		return 0, io.ErrClosedPipe
+	}
+
+	n, err := w.Write(p)
```

Comment:
missing error check here and in `Read()`? could you explain a little why we would want to return n and error if both evaluate as truthy?

Reply:
This is just a wrapper function that handles the synchronization of the underlying readers/writers so they can be swapped in a thread-safe way. We don't actually care about the return values, only that we provide them to the caller.

The exception is `Read` which needs to handle EOF a little differently so that instead of declaring we got to the end, we start reading from the next one. Similar to how `io.MultiReader` works -- but in a dynamic way.